File: src/lib/digest/text-source.ts
/**
 * Helper utilities for working with digest text sources
 */

import { promises as fs } from 'fs';
import path from 'path';
import type { Digest, FileRecordRow } from '@/types';
import { getLogger } from '@/lib/log/logger';

const log = getLogger({ module: 'DigestTextSource' });

const DATA_ROOT = process.env.MY_DATA_DIR || './data';

const EXTRA_TEXT_MIME_TYPES = new Set([
  'application/json',
  'application/xml',
  'application/javascript',
  'application/x-javascript',
  'application/x-sh',
  'application/sql',
]);

const TEXT_EXTENSIONS = new Set([
  '.md',
  '.mdx',
  '.markdown',
  '.txt',
  '.log',
  '.json',
  '.yaml',
  '.yml',
  '.csv',
  '.tsv',
]);

export type TextSourceType = 'url-digest' | 'doc-to-markdown' | 'file';

function parseJson<T>(raw: string | null): T | null {
  if (!raw) return null;
  try {
    return JSON.parse(raw) as T;
  } catch (error) {
    log.warn({ error }, 'failed to parse digest JSON payload');
    return null;
  }
}

function extractUrlCrawlMarkdown(digest: Digest | undefined): string | null {
  if (!digest?.content) return null;
  const parsed = parseJson<{ markdown?: string }>(digest.content);
  if (parsed?.markdown) {
    return parsed.markdown;
  }
  return digest.content;
}

export function getSummaryText(existingDigests: Digest[]): string | null {
  const summaryDigest =
    existingDigests.find((d) => d.digester === 'url-crawl-summary') ||
    existingDigests.find((d) => d.digester === 'summarize');

  if (!summaryDigest?.content) return null;
  const parsed = parseJson<{ summary?: string }>(summaryDigest.content);
  return parsed?.summary ?? summaryDigest.content;
}

export function getDocToMarkdown(existingDigests: Digest[]): string | null {
  const digest = existingDigests.find(
    (d) => d.digester === 'doc-to-markdown' && d.status === 'completed'
  );
  return digest?.content ?? null;
}

export function getUrlCrawlMarkdown(existingDigests: Digest[]): string | null {
  const digest = existingDigests.find(
    (d) => d.digester === 'url-crawl-content' && d.status === 'completed'
  );
  return extractUrlCrawlMarkdown(digest);
}

export function hasUrlCrawlContent(existingDigests: Digest[], minLength = 0): boolean {
  const markdown = getUrlCrawlMarkdown(existingDigests);
  return markdown ? markdown.trim().length >= minLength : false;
}

function isTextMimeType(mime: string | null): boolean {
  if (!mime) return false;
  const normalized = mime.toLowerCase();
  if (normalized.startsWith('text/')) return true;
  return EXTRA_TEXT_MIME_TYPES.has(normalized);
}

function hasTextExtension(filePath: string): boolean {
  const ext = path.extname(filePath).toLowerCase();
  return TEXT_EXTENSIONS.has(ext);
}

export function hasLocalTextContent(file: FileRecordRow, _minBytes = 0): boolean {
  if (file.is_folder) return false;
  const textLike =
    (file.mime_type ? isTextMimeType(file.mime_type) : false) ||
    hasTextExtension(file.path);
  if (!textLike) return false;
  // Allow any size - even empty files should be processed
  return true;
}

export function hasDocToMarkdownContent(existingDigests: Digest[], minLength = 0): boolean {
  const markdown = getDocToMarkdown(existingDigests);
  return markdown ? markdown.trim().length >= minLength : false;
}

export function hasAnyTextSource(
  file: FileRecordRow,
  existingDigests: Digest[],
  options?: { minUrlLength?: number; minFileBytes?: number }
): boolean {
  const urlCrawl = hasUrlCrawlContent(existingDigests, options?.minUrlLength ?? 0);
  if (urlCrawl) {
    log.info({ filePath: file.path }, 'hasAnyTextSource: TRUE from url-crawl-content');
    return true;
  }

  const docToMd = hasDocToMarkdownContent(existingDigests, options?.minUrlLength ?? 0);
  if (docToMd) {
    log.info({ filePath: file.path }, 'hasAnyTextSource: TRUE from doc-to-markdown');
    return true;
  }

  const localText = hasLocalTextContent(file, options?.minFileBytes ?? 0);
  log.info({ filePath: file.path, hasUrlCrawl: urlCrawl, hasDocToMd: docToMd, hasLocalText: localText },
    `hasAnyTextSource: ${localText ? 'TRUE from local text' : 'FALSE - no text source found'}`);
  return localText;
}

async function readLocalFile(filePath: string): Promise<string | null> {
  try {
    const value = await fs.readFile(path.join(DATA_ROOT, filePath), 'utf-8');
    return value;
  } catch (error) {
    log.warn({ filePath, error }, 'failed to read text file');
    return null;
  }
}

export async function getPrimaryTextContent(
  filePath: string,
  file: FileRecordRow,
  existingDigests: Digest[]
): Promise<{ text: string; source: TextSourceType } | null> {
  const fromUrlDigest = getUrlCrawlMarkdown(existingDigests);
  if (fromUrlDigest) {
    return { text: fromUrlDigest, source: 'url-digest' };
  }

  const fromDocDigest = getDocToMarkdown(existingDigests);
  if (fromDocDigest) {
    return { text: fromDocDigest, source: 'doc-to-markdown' };
  }

  if (!file.is_folder && hasLocalTextContent(file)) {
    const text = await readLocalFile(filePath);
    if (text) {
      return { text, source: 'file' };
    }
  }

  return null;
}

File: backend/utils/text_source.go
package utils

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"github.com/xiaoyuanzhu-com/my-life-db/config"
	"github.com/xiaoyuanzhu-com/my-life-db/db"
	"github.com/xiaoyuanzhu-com/my-life-db/log"
)

// TextSourceType represents the source of text content
type TextSourceType string

const (
	TextSourceDocToMarkdown TextSourceType = "doc-to-markdown"
	TextSourceFile          TextSourceType = "file"
)

// TextContent represents extracted text with its source
type TextContent struct {
	Text   string
	Source TextSourceType
}

// GetDocToMarkdown gets doc-to-markdown content
func GetDocToMarkdown(existingDigests []db.Digest) string {
	for _, d := range existingDigests {
		if d.Digester == "doc-to-markdown" && d.Status == "completed" && d.Content != nil {
			return *d.Content
		}
	}
	return ""
}

// GetSummaryText gets url-crawl-summary text
func GetSummaryText(existingDigests []db.Digest) string {
	for _, d := range existingDigests {
		if d.Digester == "url-crawl-summary" && d.Content != nil {
			content := *d.Content
			var parsed struct {
				Summary string `json:"summary"`
			}
			if err := json.Unmarshal([]byte(content), &parsed); err == nil && parsed.Summary != "" {
				return parsed.Summary
			}
			return content
		}
	}
	return ""
}

// IsTextFile checks if a file is a text file based on MIME type and filename
func IsTextFile(mimeType *string, filename string) bool {
	// Check by MIME type
	if mimeType != nil {
		mt := *mimeType
		if strings.HasPrefix(mt, "text/") {
			return true
		}
		textMimeTypes := []string{
			"application/json",
			"application/javascript",
			"application/xml",
			"application/x-yaml",
			"application/yaml",
		}
		for _, tm := range textMimeTypes {
			if mt == tm {
				return true
			}
		}
	}

	// Check by extension
	ext := strings.ToLower(filepath.Ext(filename))
	textExtensions := map[string]bool{
		".txt": true, ".md": true, ".markdown": true, ".json": true,
		".yaml": true, ".yml": true, ".xml": true, ".html": true,
		".htm": true, ".css": true, ".js": true, ".ts": true,
		".jsx": true, ".tsx": true, ".py": true, ".go": true,
		".rs": true, ".java": true, ".c": true, ".cpp": true,
		".h": true, ".hpp": true, ".sh": true, ".bash": true,
		".zsh": true, ".fish": true, ".sql": true, ".toml": true,
		".ini": true, ".cfg": true, ".conf": true, ".log": true,
		".env": true, ".gitignore": true, ".dockerignore": true,
	}
	return textExtensions[ext]
}

// HasLocalTextContent checks if file is a text file
func HasLocalTextContent(file *db.FileRecord) bool {
	if file.IsFolder {
		return false
	}
	return IsTextFile(file.MimeType, file.Name)
}

// HasDocToMarkdownContent checks if doc-to-markdown content exists
func HasDocToMarkdownContent(existingDigests []db.Digest, minLength int) bool {
	content := GetDocToMarkdown(existingDigests)
	return len(strings.TrimSpace(content)) >= minLength
}

// HasAnyTextSource checks if any text source is available
func HasAnyTextSource(file *db.FileRecord, existingDigests []db.Digest, minLength int) bool {
	if HasDocToMarkdownContent(existingDigests, minLength) {
		return true
	}
	return HasLocalTextContent(file)
}

// ReadLocalFile reads a text file from data directory
func ReadLocalFile(filePath string) (string, error) {
	cfg := config.Get()
	fullPath := filepath.Join(cfg.UserDataDir, filePath)

	data, err := os.ReadFile(fullPath)
	if err != nil {
		log.Warn().Str("path", filePath).Err(err).Msg("failed to read text file")
		return "", err
	}

	return string(data), nil
}

// GetPrimaryTextContent gets the primary text content for a file with priority order:
// 1. Document to markdown (for PDFs, DOCX, etc.)
// 2. Local file content (for text files)
func GetPrimaryTextContent(filePath string, file *db.FileRecord, existingDigests []db.Digest) *TextContent {
	// 1. Document to markdown (for PDFs, DOCX, etc.)
	if text := GetDocToMarkdown(existingDigests); text != "" {
		return &TextContent{Text: text, Source: TextSourceDocToMarkdown}
	}

	// 2. Local file content (for text files)
	if !file.IsFolder && HasLocalTextContent(file) {
		text, err := ReadLocalFile(filePath)
		if err == nil && text != "" {
			return &TextContent{Text: text, Source: TextSourceFile}
		}
	}

	return nil
}
